File: solutions/clojure/pig-latin/1/src/pig_latin.clj
(ns pig-latin
  (:require [clojure.string :as s]))

(def exceptions #"^([aeiou]|yt|xr).*")

(def lead-consonants #"([^aeiou]*qu|[^aeiou]*)([aeiou].*)")

(defn- pig-word [w]
  (if (re-matches exceptions w)
    (str w "ay")
    (if-let [[_ head tail] (re-matches lead-consonants w)]
      (str tail head "ay")
      (str w "ay"))))

(defn translate [phrase]
  (->>
    (s/split phrase #"\s")
    (map pig-word)
    (s/join " ")))

File: solutions/elixir/pig-latin/1/lib/pig_latin.ex
defmodule PigLatin do
  @vowels ~w(a e i o u y)

  @doc """
  Given a `phrase`, translate it a word at a time to Pig Latin.
  """
  @spec translate(phrase :: String.t()) :: String.t()
  def translate(phrase) do
    phrase
    |> String.split(" ")
    |> Enum.map_join(" ", fn chunk ->
      chunk
      |> String.graphemes()
      |> Enum.split_while(&(&1 not in @vowels))
      |> to_latin()
      |> Enum.join()
    end)
  end

  defp to_latin({[], ["y", ch | _] = v}) when ch in @vowels, do: [tl(v), "yay"]
  defp to_latin({c, ["y" | _] = v}), do: [v, c, "ay"]
  defp to_latin({["x", "r"], v}), do: ["xr", v, "ay"]

  defp to_latin({c, ["u" | v]}) do
    if List.last(c) == "q" do
      [v, c, "uay"]
    else
      ["u", v, c, "ay"]
    end
  end

  defp to_latin({c, v}), do: [v, c, "ay"]
end
